### PR TITLE
Fix GetComputeRunningProcesses on CUDA 10.x

### LINKED ISSUE #21: Wrong output of `device.GetComputeRunningProcesses()` given multiple processes
Hi Kevin & Evan,

When multiple processes run on one GPU, I found the output of `device.GetComputeRunningProcesses()` is wrong -- the values are misplaced across different process's `ProcessInfo`. The bug appears on both V100-SXM2-16GB and GTX 1080 Ti, with CUDA version 10.2.

The testing code snippet is as follows.
```go
package main

import "fmt"
import "github.com/NVIDIA/go-nvml/pkg/nvml"

func main() {
	nvml.Init()
	device, _ := nvml.DeviceGetHandleByIndex(0)
	processInfos, _ := device.GetComputeRunningProcesses()
	for i, processInfo := range processInfos {
		fmt.Printf("\t[%2d] ProcessInfo: %v\n", i, processInfo)
	}
}
```

On V100 machines, I got this.
```bash
$nvidia-smi -L
GPU 0: Tesla V100-SXM2-16GB (UUID: GPU-?) # UUID manually removed

$nvidia-smi
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 440.118.02   Driver Version: 440.118.02   CUDA Version: 10.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  On   | 00000000:00:08.0 Off |                    0 |
| N/A   55C    P0   130W / 300W |  13244MiB / 16160MiB |     53%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0     48959      C   python                                      1179MiB |
|    0     72754      C   python                                      9611MiB |
|    0     73422      C   python                                      2443MiB |
+-----------------------------------------------------------------------------+

$go run main.go
	[ 0] ProcessInfo: {72754 10077863936 73422 0}
	[ 1] ProcessInfo: {2561671168 48959 1236271104 0}
	[ 2] ProcessInfo: {0 0 0 0}
# it is expected to be
#	[ 0] ProcessInfo: {72754 10077863936 0 0} # {PID, 9611 MiB, 0, 0}
#	[ 1] ProcessInfo: {73422 2561671168 0 0}  # {PID, 2443 MiB, 0, 0}
#	[ 2] ProcessInfo: {48959 1236271104 0 0}  # {PID, 1179 MiB, 0, 0}
```

On GTX 1080 Ti machines, I got this.
```bash
$nvidia-smi -L
GPU 0: GeForce GTX 1080 Ti (UUID: GPU-?)
GPU 1: GeForce GTX 1080 Ti (UUID: GPU-?)
GPU 2: GeForce GTX 1080 Ti (UUID: GPU-?)
GPU 3: GeForce GTX 1080 Ti (UUID: GPU-?)
GPU 4: GeForce GTX 1080 Ti (UUID: GPU-?)
GPU 5: GeForce GTX 1080 Ti (UUID: GPU-?)
GPU 6: GeForce GTX 1080 Ti (UUID: GPU-?)
GPU 7: GeForce GTX 1080 Ti (UUID: GPU-?) # UUID manually removed

$nvidia-smi -i 0
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 440.64       Driver Version: 440.64       CUDA Version: 10.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 108...  On   | 00000000:3D:00.0 Off |                  N/A |
| 38%   67C    P2   247W / 250W |   6630MiB / 11178MiB |    100%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0     25907      C   python                                      2581MiB |
|    0     27576      C   python                                      4039MiB |
+-----------------------------------------------------------------------------+

$go run main.go
	[ 0] ProcessInfo: {25907 2706374656 27576 0}
	[ 1] ProcessInfo: {4235198464 0 0 0}
# it is expected to be
#	[ 0] ProcessInfo: {25907 2706374656 0 0} # {PID, 2581 MiB, 0, 0}
#	[ 1] ProcessInfo: {27576 4235198464 0 0} # {PID, 4039 MiB, 0, 0}
```

As a quick fix, I write a wrapper function to correct the faulty processInfo as it returns, hoping this bug could be solved in the near future.

After all, thanks for providing such nice library, especially the useful "thread-safe" feature. :)

### PR DESCRIPTION
**Pull request by XuehaiPan**
The definition of `struct nvmlProcessInfo_st` has introduce two new fields `gpuInstanceId` and `computeInstanceId` in newer NVIDIA drivers (CUDA 11.x). And this changes will break the backward compatibility for old NVIDIA drivers (CUDA 10.x).

This PR fixes issue #21 for `nvmlDeviceGetComputeRunningProcesses` `nvmlDeviceGetGraphicsRunningProcesses` on CUDA 10.x caused by the data structure size change. It simply calls the v1 version function (feed with v1 data structure) and convert the results into v2 ones on CUDA 10.x. And for CUDA 11.x, call the v2 version directly.

There is another solution (PR #22) that calls the v1 version (but feed with v2 data structure) and correct the results with some adjustments on CUDA 10.x. As commented in https://github.com/NVIDIA/go-nvml/issues/21#issuecomment-897657714, this implementation should cover a large number of configuration situations (different CPU bit width / CPU arch / complie options, etc.).

---

The results on CUDA 10.x

```console
$ python3 -c 'import time; import cupy as cp; x = cp.zeros((1, 1)); time.sleep(120)' &
[1] 5404
$ python3 -c 'import time; import cupy as cp; x = cp.zeros((1, 1)); time.sleep(120)' &
[2] 5456
$ python3 -c 'import time; import cupy as cp; x = cp.zeros((1, 1)); time.sleep(120)' &
[3] 5481
$ python3 -c 'import time; import cupy as cp; x = cp.zeros((1, 1)); time.sleep(120)' &
[4] 5504
$ python3 -c 'import time; import cupy as cp; x = cp.zeros((1, 1)); time.sleep(120)' &
[5] 5529

$ go run go-nvml/examples/compute-processes
Found 5 processes on device 0
        [ 0] ProcessInfo: {Pid:5404 UsedGpuMemory:173015040 GpuInstanceId:4294967295 ComputeInstanceId:4294967295}
        [ 1] ProcessInfo: {Pid:5456 UsedGpuMemory:173015040 GpuInstanceId:4294967295 ComputeInstanceId:4294967295}
        [ 2] ProcessInfo: {Pid:5481 UsedGpuMemory:173015040 GpuInstanceId:4294967295 ComputeInstanceId:4294967295}
        [ 3] ProcessInfo: {Pid:5504 UsedGpuMemory:173015040 GpuInstanceId:4294967295 ComputeInstanceId:4294967295}
        [ 4] ProcessInfo: {Pid:5529 UsedGpuMemory:173015040 GpuInstanceId:4294967295 ComputeInstanceId:4294967295}
Found 0 processes on device 1
Found 0 processes on device 2
```

Fixes #21
Closes #22

### CODE REVIEW

File: gen/nvml/device.go
Comment (elezar) [inline gen/nvml/device.go:945]: I think we would need to use `InfoCount` as the loop variable here as this is modified in the call above to return the actual number of elements.

File: gen/nvml/device.go
Comment (elezar) [inline gen/nvml/device.go:986]: This should be `usesNvmlDeviceGetGraphicsRunningProcesses_v1`.

**Review by elezar — CHANGES_REQUESTED** [review]
Thanks a lot for the submission @XuehaiPan. This is definitely more robust than what I had implemented.

I have left some comments below. There are one or two that need to be addressed before we can merge this.

File: gen/nvml/init.go
Comment (XuehaiPan) [reply gen/nvml/init.go:97]: Okay, it seems more readable.

File: gen/nvml/init.go
Comment (XuehaiPan) [reply gen/nvml/init.go:100]: As the `Makefile` defines:

https://github.com/NVIDIA/go-nvml/blob/10a3a255c928e24f42c3d43ed6f0a087ece02e7a/Makefile#L86-L96

We copy the `nvml.h` from a docker container. If we want to add `ProcessInfo_v1` struct definition to `nvml.h`, we have to add it manually on each time header update.

Comment (elezar) [reply gen/nvml/init.go:100]: OK. I'll add that to a list of todo's for myself instead. I don't see this as blocking the PR from merging.

File: gen/nvml/device.go
Comment (XuehaiPan) [reply gen/nvml/device.go:940]: I think it seems too cumbersome to split this into two almost identical functions. And `deviceGetComputeRunningProcesses_v1` returns the result of type `ProcessInfo_v2` which is a bit misleading.

File: gen/nvml/device.go
Comment (XuehaiPan) [reply gen/nvml/device.go:986]: Thanks. I'll fix this right away.

File: gen/nvml/device.go
Comment (XuehaiPan) [reply gen/nvml/device.go:945]: Thanks. I'll fix this right away.

File: gen/nvml/device.go
Comment (elezar) [reply gen/nvml/device.go:940]: OK. No problem.

**Review by elezar — APPROVED** [review]
Looks good @XuehaiPan.

Thanks for the work on this!

**Comment by klueska** [comment]
I think this is fine to unblock you for now, but I wouldn't spend too much time "perfecting" it as we will need tocome up with a more general solution in the near future.

I see the "real" solution as:
1. Generate a `types.go` for each new version of NVML we support and put it in its own package under e.g. `pkg/types/v1.11.1`.
2. For any type we care about making sure is usable in an array context, we need to have a helper function that:
a. Dynamically checks for the current version of NVML on the host.
b. Passes the correct (versioned) type to the underyling call that returns the array
c. Walks the array and copies each entry into the "latest" type (i.e. the one that may have extended fields for newer versions)
d. Returns an array of the latest type

**Comment by elezar** [comment]
Thanks for the input @klueska. It was also pointed out to me that more recent versions of the `nvml.h` file has an explicit `_v1` type. As such the current workaround should be simpler then, but we would still have to ensure that the returned structs have the right types.

**Pull request merged**
